### PR TITLE
Define a start method

### PR DESCRIPTION
**Pull request by jcbeyler**
Adding the definition of a start method that would be a top-level module statement.

### CODE REVIEW

**Comment by jfbastien** [comment]
IIUC this isn't the same thing as what @lukewagner suggested in https://github.com/WebAssembly/design/issues/398#issuecomment-148080824 ?

**Comment by jcbeyler** [comment]
Yes exactly, he asked me to do a PR, so here it is: https://github.com/WebAssembly/design/issues/398#issuecomment-151610568

**Comment by lukewagner** [comment]
lgtm, thanks!

**Comment by jfbastien** [comment]
I expected something more concrete, mentioning `(start $func)` or something of the kind.

**Comment by jcbeyler** [comment]
Added an example for JF

File: Modules.md [hunk 1]
```diff
@@ -122,6 +124,30 @@ shared `malloc` and coordinated global address ranges). Instead, the
...
+* The function can also be exported
+* There can only be at most one start node per module
+
+For example, a start node in a module will be:
+
+```(start $start_function)```
+
+or
+
+```(start 0)```
```
Comment (jfbastien) [inline Modules.md]: Could you explain what this would do?

**Comment by jfbastien** [comment]
lgtm with small explanation. Thanks!

File: Modules.md (hunk 1 above)
Comment (jcbeyler) [inline Modules.md]: done I think :)

**Comment by dschuff** [comment]
We've started some discussion about how this should integrate with emscripten in https://github.com/kripken/emscripten/issues/4218

It's not really clear from the current text in the design repo how it should work with the web embedding. e.g. when you instantiate the module, is the VM expected to call the start function automatically (if so, when? before `Wasm.instantiateModule` returns?), or should that be done by the enclosing JS code?

I think it should be put under the control of the JS code. For example (as a strawman) the object returned by `Wasm.instantiateModule` could have a way to determine which function is the start function, and then  the JS would have to call that function before calling anything else on the module. Or just have a method with the same name on every instance that is called by the JS, which itself calls the start function. Or whatever. The point is that it would allow the JS to instantiate the module, then modify the linear memory and do whatever JS wrapper stuff it would need to, then call the wasm start function (which would presumably then call its own global initializer functions).

**Comment by dschuff** [comment]
oh oops, I meant to post this in https://github.com/WebAssembly/design/issues/398, not in this PR.

**Pull request merged**